src/gluonts/itertools.py

```diff
@@ -129,18 +149,79 @@ def __len__(self) -> int:
         return len(self.iterable)
 
 
-class IterableSlice(Iterable):
+# can't make this a dataclass because of pytorch-lightning assumptions
+class IterableSlice:
     """
     An iterable version of `itertools.islice`, i.e. one that can be iterated
-    over multiple times.
+    over multiple times:
+
+        >>> isl = IterableSlice([1, 2, 3, 4, 5], 3)
+        >>> list(isl)
+        [1, 2, 3]
+        >>> list(isl)
+        [4, 5]
+        >>> list(isl)
+        []
+
+    This needs to be a class to support re-entry iteration.
     """
 
-    def __init__(self, iterable: Iterable, length: Optional[int]) -> None:
+    def __init__(self, iterable, length):
         self.iterable = iterable
         self.length = length
+        self.next()
+
+    def next(self):
+        self.current = list(itertools.islice(self.iterable, self.length))
 
     def __iter__(self):
-        return itertools.islice(self.iterable, self.length)
+        yield from self.current
+        self.next()
 
     def __len__(self) -> int:
-        return len(self.iterable)
```

Comment:
So this was wrong?

Reply:
Right, I'm wondering if it makes even sense to have `__len__` here. Also, we probably lacked tests for it.

Reply:
I'm not so sure either, I guess I can remove it and check what happens?

Reply:
Yes, let's do that